### PR TITLE
feat: multithreaded bulk import

### PR DESCRIPTION
**Pull request by tamassoltesz**
## Summary of change

Core implementation of Bulk User Migration.

## Related issues

- [Bulk User Import issue](https://github.com/supertokens/supertokens-core/issues/912)

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your
changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here
highlighting the necessary changes)

## Checklist for important updates

- [x] Changelog has been updated
    - [x] If there are any db schema changes, mention those changes clearly
- [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [ ] `pluginInterfaceSupported.json` file has been updated (if needed)
- [x] Changes to the version if needed
    - In `build.gradle`
- [ ] If added a new paid feature, edit the `getPaidFeatureStats` function in FeatureFlag.java file
- [x] Had installed and ran the pre-commit hook
- [ ] If there are new dependencies that have been added in `build.gradle`, please make sure to add them
  in `implementationDependencies.json`.
- [ ] Update function `getValidFields` in `io/supertokens/config/CoreConfig.java` if new aliases were added for any core
  config (similar to the `access_token_signing_key_update_interval` config alias).
- [ ] Issue this PR against the latest non released version branch.
    - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the
      latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    - If no such branch exists, then create one from the latest released branch.
- [ ] If added a foreign key constraint on `app_id_to_user_id` table, make sure to delete from this table when deleting
  the user as well if `deleteUserIdMappingToo` is false.

## Remaining TODOs for this PR

- [ ] Item1
- [ ] Item2

### CODE REVIEW

**Review by sattvikc — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/bulkimport/BulkImportBackgroundJobManager.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    public static BULK_IMPORT_BACKGROUND_PROCESS_STATUS startBackgroundJob(Main main, Integer batchSize) throws TenantOrAppNotFoundException {
+        ProcessBulkImportUsers processBulkImportUsersCron = (ProcessBulkImportUsers) main.getResourceDistributor().getResource(new TenantIdentifier(null, null, null), ProcessBulkImportUsers.RESOURCE_KEY);
+        processBulkImportUsersCron.setBatchSize(batchSize);
+        Cronjobs.addCronjob(main, processBulkImportUsersCron);
+        return BULK_IMPORT_BACKGROUND_PROCESS_STATUS.ACTIVE;
+    }
+
+    public static BULK_IMPORT_BACKGROUND_PROCESS_STATUS stopBackgroundJob(Main main) throws TenantOrAppNotFoundException {
+        ProcessBulkImportUsers processBulkImportUsersCron = (ProcessBulkImportUsers) main.getResourceDistributor().getResource(new TenantIdentifier(null, null, null), ProcessBulkImportUsers.RESOURCE_KEY);
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/bulkimport/BulkImportBackgroundJobManager.java:35]: add test if possible

File: src/test/java/io/supertokens/test/bulkimport/apis/BulkImportBackgroundJobManagerAPITest.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        return HttpRequestForTesting.sendJsonPOSTRequest(main, "",
+                BACKGROUNDJOB_MANAGER_ENDPOINT,
+                request, 1000, 10000, null, Utils.getCdiVersionStringLatestForTests(), null);
+    }
+
+    private static JsonObject loadBulkImportCronjobStatus(Main main) throws HttpResponseException, IOException {
+        return HttpRequestForTesting.sendGETRequest(main, "", BACKGROUNDJOB_MANAGER_ENDPOINT, null,
+                1000, 1000, null, Utils.getCdiVersionStringLatestForTests(), null);
+    }
+
```
Comment (sattvikc) [inline src/test/java/io/supertokens/test/bulkimport/apis/BulkImportBackgroundJobManagerAPITest.java:122]: What's the behaviour if the job is stopped when an import is happening? can we test that as well?

File: src/main/java/io/supertokens/webserver/api/bulkimport/BulkImportAPI.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+            throw new ServletException(new BadRequestException("This API is not supported in the in-memory database."));
+        }
+
+        String statusString = InputParser.getQueryParamOrThrowError(req, "status", true);
+        String paginationToken = InputParser.getQueryParamOrThrowError(req, "paginationToken", true);
+        Integer limit = InputParser.getIntQueryParamOrThrowError(req, "limit", true);
+
+        if (limit != null) {
+            if (limit > BulkImport.GET_USERS_PAGINATION_MAX_LIMIT) {
+                throw new ServletException(
```
Comment (sattvikc) [inline src/main/java/io/supertokens/webserver/api/bulkimport/BulkImportAPI.java:73]: Are u aware of typical post body size if say we are adding 10000 users? it would be a useful information to be able to set max post body size in the nginx for example.

File: src/main/java/io/supertokens/webserver/api/bulkimport/BulkImportBackgroundJobManagerAPI.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    public BulkImportBackgroundJobManagerAPI(Main main) {
+        super(main, "");
+    }
+
+    @Override
+    public String getPath() {
+        return "/bulk-import/backgroundjob";
+    }
+
+    //TODO what's stopping two independent calls interfering with each-other?
```
Comment (sattvikc) [inline src/main/java/io/supertokens/webserver/api/bulkimport/BulkImportBackgroundJobManagerAPI.java:48]: TODO

File: src/main/java/io/supertokens/webserver/api/bulkimport/CountBulkImportUsersAPI.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import io.supertokens.storageLayer.StorageLayer;
+import io.supertokens.webserver.InputParser;
+import io.supertokens.webserver.WebserverAPI;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+
+public class CountBulkImportUsersAPI extends WebserverAPI {
+  public CountBulkImportUsersAPI(Main main) {
+    super(main, "");
```
Comment (sattvikc) [inline src/main/java/io/supertokens/webserver/api/bulkimport/CountBulkImportUsersAPI.java:40]: we could have an rid for bulk import. you could add it to all the bulk import APIs

File: src/main/java/io/supertokens/webserver/api/bulkimport/ImportUserAPI.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+            AuthRecipeUserInfo importedUser = BulkImport.importUser(main, appIdentifier, user);
+
+            JsonObject result = new JsonObject();
+            result.addProperty("status", "OK");
+            result.add("user", importedUser.toJson());
+            super.sendJsonResponse(200, result, resp);
+        } catch (StorageQueryException e) {
+            JsonArray errors = new JsonArray();
+            if(e.getCause() instanceof BulkImportBatchInsertException){
```
Comment (sattvikc) [inline src/main/java/io/supertokens/webserver/api/bulkimport/ImportUserAPI.java:92]: this exception should be caught and thrown in BulkImport.importUser

File: src/main/java/io/supertokens/utils/Utils.java [hunk 7]
```diff
@@ -432,6 +434,12 @@ public static JsonElement toJsonTreeWithNulls(Object src) {
         return new GsonBuilder().serializeNulls().create().toJsonTree(src);
     }
 
+
+    public static boolean isAccountLinkingEnabled(Main main, AppIdentifier appIdentifier) throws StorageQueryException, TenantOrAppNotFoundException {
+        return Arrays.stream(FeatureFlag.getInstance(main, appIdentifier).getEnabledFeatures())
+                .anyMatch(t -> t == EE_FEATURES.ACCOUNT_LINKING || t == EE_FEATURES.MFA);
+    }
```
Comment (sattvikc) [inline src/main/java/io/supertokens/utils/Utils.java:441]: not sure why this change appears here

**Review by sattvikc — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/bulkimport/BulkImportBackgroundJobManager.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+import io.supertokens.Main;
+import io.supertokens.cronjobs.Cronjobs;
+import io.supertokens.cronjobs.bulkimport.ProcessBulkImportUsers;
+import io.supertokens.pluginInterface.multitenancy.TenantIdentifier;
+import io.supertokens.pluginInterface.multitenancy.exceptions.TenantOrAppNotFoundException;
+
+public class BulkImportBackgroundJobManager {
+
+    public static BULK_IMPORT_BACKGROUND_PROCESS_STATUS startBackgroundJob(Main main, Integer batchSize) throws TenantOrAppNotFoundException {
```
[the remark this answers is not in the document]
Comment (tamassoltesz) [reply src/main/java/io/supertokens/bulkimport/BulkImportBackgroundJobManager.java:27]: no

File: src/main/java/io/supertokens/bulkimport/BulkImport.java [hunk 9]
```diff
@@ -0,0 +1,815 @@
...
+    // Maximum number of users to return in a single page when calling /bulk-import/users GET
+    public static final int GET_USERS_PAGINATION_MAX_LIMIT = 500;
+    // Default number of users to return when no specific limit is given in /bulk-import/users GET
+    public static final int GET_USERS_DEFAULT_LIMIT = 100;
+    // Maximum number of users that can be deleted in a single operation
+    public static final int DELETE_USERS_MAX_LIMIT = 500;
+    // Number of users to process in a single batch of ProcessBulkImportUsers Cron Job
+    public static final int PROCESS_USERS_BATCH_SIZE = 10000;
+    // Time interval in seconds between two consecutive runs of ProcessBulkImportUsers Cron Job
+    public static final int PROCESS_USERS_INTERVAL_SECONDS = 1;
```
[the remark this answers is not in the document]
Comment (tamassoltesz) [reply src/main/java/io/supertokens/bulkimport/BulkImport.java:100]: yes. We want to squeeze out as much performance as we can. Setting it to 1 sec means that the new round of batch import processing is going to happen 1 seconds after the current round finished. This is the reason why the start/stop cronjob api was done: to not waste the resources when we don't need to import anything.

File: src/main/java/io/supertokens/config/CoreConfig.java [hunk 10]
```diff
@@ -772,6 +782,10 @@ void normalizeAndValidate(Main main, boolean includeConfigFilePath) throws Inval
             }
         }
 
+        if (bulk_migration_parallelism < 1) {
```
[the remark this answers is not in the document]
Comment (tamassoltesz) [reply src/main/java/io/supertokens/config/CoreConfig.java:785]: theoretically speeking the upper end is limitless, purely depending on the hardware. Of course, if this is set to an incredibely high number without the supporting hardware architecture beneath it, the JVM _can_ crash when actually trying to bulk import. (The number of threads that are defined here are only instantiated when a bulk import batch is loaded for processing)

File: src/main/java/io/supertokens/webserver/api/bulkimport/BulkImportBackgroundJobManagerAPI.java (hunk 4 above)
Comment (tamassoltesz) [reply src/main/java/io/supertokens/webserver/api/bulkimport/BulkImportBackgroundJobManagerAPI.java:48]: yeah, but I don't have an answer for this. I can remove the comment but I don't have a solution for the written problem, apart from the clients who are orchestrating the imports should be aware. So, should I remove the comment? Do you have a better idea for solving this?

File: src/main/java/io/supertokens/utils/Utils.java (hunk 7 above)
Comment (tamassoltesz) [reply src/main/java/io/supertokens/utils/Utils.java:441]: Because this was moved from `AuthRecipe.java `to the `Utils.java` as it was used in `BulkImportUserUtils` too, but with code duplication. I removed the code duplication and extracted this to a place where both participants can reach and doesn't go against the dependencies between the code layers.

File: src/test/java/io/supertokens/test/bulkimport/apis/BulkImportBackgroundJobManagerAPITest.java (hunk 2 above)
Comment (tamassoltesz) [reply src/test/java/io/supertokens/test/bulkimport/apis/BulkImportBackgroundJobManagerAPITest.java:122]: Processing is stopped. Added a new test for that.

File: src/main/java/io/supertokens/webserver/api/bulkimport/BulkImportAPI.java (hunk 3 above)
Comment (tamassoltesz) [reply src/main/java/io/supertokens/webserver/api/bulkimport/BulkImportAPI.java:73]: well, 10k randomized users with multiple LMs and so on, are around 7.5MB (written to a json file). I don't know if that counts as typical or not.

File: src/main/java/io/supertokens/webserver/api/bulkimport/CountBulkImportUsersAPI.java (hunk 5 above)
Comment (tamassoltesz) [reply src/main/java/io/supertokens/webserver/api/bulkimport/CountBulkImportUsersAPI.java]: done

File: src/main/java/io/supertokens/webserver/api/bulkimport/ImportUserAPI.java (hunk 6 above)
Comment (tamassoltesz) [reply src/main/java/io/supertokens/webserver/api/bulkimport/ImportUserAPI.java]: done

File: src/main/java/io/supertokens/cronjobs/Cronjobs.java [hunk 11]
```diff
@@ -100,6 +100,17 @@ public static void addCronjob(Main main, CronTask task) {
         }
     }
 
+    // TODO test for this
```
[the remark this answers is not in the document]
Comment (tamassoltesz) [reply src/main/java/io/supertokens/cronjobs/Cronjobs.java]: done

File: src/main/java/io/supertokens/bulkimport/BulkImportBackgroundJobManager.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (tamassoltesz) [reply src/main/java/io/supertokens/bulkimport/BulkImportBackgroundJobManager.java:35]: nothing. added a test for that

File: devConfig.yaml [hunk 12]
```diff
@@ -170,3 +170,8 @@ disable_telemetry: true
 
 # (Optional | Default: null) string value. The encryption key used for saving OAuth client secret on the database.
 # oauth_client_secret_encryption_key:
+
+# (DIFFERENT_ACROSS_TENANTS | OPTIONAL | Default: number of available processor cores) int value. If specified,
```
[the remark this answers is not in the document]
Comment (tamassoltesz) [reply devConfig.yaml]: changed

File: config.yaml [hunk 13]
```diff
@@ -170,3 +170,7 @@ core_config_version: 0
 
 # (Optional | Default: null) string value. The encryption key used for saving OAuth client secret on the database.
 # oauth_client_secret_encryption_key:
+
+# (DIFFERENT_ACROSS_TENANTS | OPTIONAL | Default: number of available processor cores) int value. If specified,
+# the supertokens core will use the specified number of threads to complete the migration of users.
+# bulk_migration_parallelism:
```
[the remark this answers is not in the document]
Comment (tamassoltesz) [reply config.yaml:176]: done

File: src/main/java/io/supertokens/webserver/api/bulkimport/BulkImportBackgroundJobManagerAPI.java (hunk 4 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/bulkimport/BulkImportBackgroundJobManagerAPI.java:48]: maybe `startBackgroundJob` and`stopBackgroundJob` should use a shared lock?

File: src/main/java/io/supertokens/webserver/api/bulkimport/BulkImportAPI.java (hunk 3 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/bulkimport/BulkImportAPI.java:72]: we just need to notify the SaaS team as to how much max post body size should be set for this endpoint specifically.

Comment (tamassoltesz) [reply src/main/java/io/supertokens/webserver/api/bulkimport/BulkImportAPI.java:72]: well, I don't have a better number. It's around 7.5MB

**Review by sattvikc — APPROVED** [review]
